// @shopify/polaris/esnext/components/shared.js
export const scrollable = {
    props: { 'data-polaris-scrollable': true },
    selector: '[data-polaris-scrollable]',
};
export const overlay = {
    props: { 'data-polaris-overlay': true },
    selector: '[data-polaris-overlay]',
};
export const layer = {
    props: { 'data-polaris-layer': true },
    selector: '[data-polaris-layer]',
};
export const unstyled = {
    props: { 'data-polaris-unstyled': true },
    selector: '[data-polaris-unstyled]',
};
export const dataPolarisTopBar = {
    props: { 'data-polaris-top-bar': true },
    selector: '[data-polaris-top-bar]',
};
export const headerCell = {
    props: { 'data-polaris-header-cell': true },
    selector: '[data-polaris-header-cell]',
};
export const DATA_ATTRIBUTE = {
    overlay,
    layer,
};
// these match our values in duration.scss
export var Duration;
(function (Duration) {
    Duration[Duration["Instant"] = 0] = "Instant";
    Duration[Duration["Fast"] = 100] = "Fast";
    Duration[Duration["Base"] = 200] = "Base";
    Duration[Duration["Slow"] = 300] = "Slow";
    Duration[Duration["Slower"] = 400] = "Slower";
    Duration[Duration["Slowest"] = 500] = "Slowest";
})(Duration || (Duration = {}));
