// @shopify/polaris/esnext/components/Navigation/components/UserMenu/UserMenu.js
import * as React from 'react';
import { UserMenuModifier } from '../../../TopBar';
/** @deprecated Use <TopBar.UserMenu /> instead. */
class UserMenu extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            open: false,
        };
        this.handleToggle = () => {
            const { open } = this.state;
            this.setState({ open: !open });
        };
        // eslint-disable-next-line no-console
        console.warn('Deprecation: <Navigation.UserMenu /> is deprecated and will be removed in the next major version. Use <TopBar.UserMenu /> instead.');
    }
    render() {
        const { name, detail, actions, message, avatarInitials, avatarSource, } = this.props;
        const { open } = this.state;
        const userMenuProps = {
            actions: actions || [],
            message,
            name: name || '',
            detail,
            initials: avatarInitials,
            avatar: avatarSource,
            onToggle: this.handleToggle,
            open,
        };
        return <UserMenuModifier userMenuProps={userMenuProps}/>;
    }
}
export default UserMenu;
